### PR TITLE
Add routing

### PR DESCRIPTION
**Pull request by bohdan-meltiukhov**
# Routing
This PR adds a new page and allows switching between the pages using routing. The PR includes:
- The registration page;
- The configured ESLint, which rules are applied to the project;
- The router with tests;
- The basic validation, which is improved in the `services` branch;
- The page 404 for routes that do not exist.

### CODE REVIEW

File: ui-components/app/components/login-form-footer/index.js [hunk 1]
```diff
@@ -1,57 +1,99 @@
...
+   */
+  initNestedComponents() {
+    const buttonContainer = this.rootElement.querySelector('.submit-button');
+    this.button = new Button(buttonContainer, {
+      buttonText: this._buttonText,
+    });
+  }
+
+  /**
+   * Provides a new button click handler to the button.
```
Comment (dkalinin) [inline ui-components/app/components/login-form-footer/index.js]: Adds a new button click handler.

File: ui-components/app/components/login-form-footer/index.js [hunk 2]
```diff
@@ -1,57 +1,99 @@
...
+   *
+   * @param {Function} handler - The function to be called when the form button is pressed.
+   */
+  addButtonClickHandler(handler) {
+    this.button.addClickHandler(handler);
+  }
+
+  /**
+   * Assign the provided button text to the private field and changes the text of the button.
```
Comment (dkalinin) [inline ui-components/app/components/login-form-footer/index.js]: Please remove information about internal (private) implementation.

File: ui-components/router.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  /**
+   * Creates an instance of the router with set properties.
+   *
+   * @param {RouterProperties} properties - The properties for the current router.
+   */
+  constructor(properties) {
+    this._rootElement = properties.rootElement;
+    this._pageMapping = properties.pageMapping;
+    this._defaultLocation = properties.defaultLocation;
```
Comment (dkalinin) [inline ui-components/router.js]: Please add 404 page configuration.

File: ui-components/tests/router/index.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    'default location if the hash is changed to empty.');
+});
+
+test('should change the pages when the location hash changes.', (assert) => {
+  const windowMock = new WindowMock();
+
+  const rootElement = document.createElement('div');
+
+  const properties = {
+    rootElement: rootElement,
```
Comment (dkalinin) [inline ui-components/tests/router/index.js]: rootElement: rootElement, ==> rootElement,

File: ui-components/app/components/validator.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+   */
+  validate(formInput, properties) {
+    const inputValue = formInput.inputValue;
+    const errorMessages = [];
+    let anyErrorFound = false;
+
+    if (properties.minLength && inputValue.length < properties.minLength) {
+      errorMessages.push(`have ${properties.minLength} or more characters`);
+      anyErrorFound = true;
+    }
```
Comment (dkalinin) [inline ui-components/app/components/validator.js]: Let's refactor this according to our discussion.

File: ui-components/app/components/login-form-footer/index.js (hunk 1 above)
Comment (bohdan-meltiukhov) [reply ui-components/app/components/login-form-footer/index.js]: I changed the JSDoc now.

File: ui-components/app/components/login-form-footer/index.js (hunk 2 above)
Comment (bohdan-meltiukhov) [reply ui-components/app/components/login-form-footer/index.js]: Done.

File: ui-components/tests/router/index.js (hunk 4 above)
Comment (bohdan-meltiukhov) [reply ui-components/tests/router/index.js]: Done.

File: ui-components/app/components/validator.js (hunk 5 above)
Comment (bohdan-meltiukhov) [reply ui-components/app/components/validator.js]: Refactored in the PR 'Services'.

File: ui-components/router.js (hunk 3 above)
Comment (bohdan-meltiukhov) [reply ui-components/router.js]: Done.

File: ui-components/app/components/not-found/index.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+  /** @inheritdoc */
+  markup() {
+    return `
+        <div class="application-box form-dialog">
+            <img src="app/images/logo.png" class="logo" alt="logo">
+            
+            <div class="form-header"></div>
+            
+            <p>Not found</p>
```
Comment (dkalinin) [inline ui-components/app/components/not-found/index.js:27]: Let's have a link to '#'.

File: ui-components/router.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
+/**
+ * The class that allows switching between different pages.
+ */
+export default class Router {
+  /**
+   * The class for providing the router properties via the constructor.
+   *
+   * @typedef RouterProperties
+   * @property {Element} rootElement - The parent element for the router.
+   * @property {object} pageMapping - The object that defines all the possible routes.
```
Comment (dkalinin) [inline ui-components/router.js:10]: Please specify key and value types of page mapping.

File: ui-components/router.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
+/**
```
Comment (dkalinin) [inline ui-components/router.js:1]: Please move this file to app folder.

File: ui-components/router.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  /**
+   * Checks the location hash and loads the required page.
+   */
+  init() {
+    if (this._window.location.hash === '') {
+      this._window.location.hash = `#${this._defaultLocation}`;
+    }
+
+    const initialHash = this._window.location.hash.slice(1);
+    this.loadPage(initialHash);
```
Comment (dkalinin) [inline ui-components/router.js:39]: - [ ] Move duplicated code to a private method. Use this method in `hashchange` handler.

**Pull request merged**